net/xia/ppal_ether/main.c

```diff
@@ -222,6 +223,69 @@ static void detach_neigh_to_interface(struct fib_xid_ether_main *mether)
 	ether_interface_put(einterface);
 }
 
+/* Creates a cached hdr and alocates to a mfxid. */
+static int xia_ether_header_cache(const struct fib_xid_ether_main *mfxid,
+				  struct hh_cache *hh, __be16 type)
+{
+	struct ethhdr *eth;
+	const struct net_device *dev = mfxid->host_interface;
+
+	eth = (struct ethhdr *)
+	    (((u8 *)hh->hh_data) + (HH_DATA_OFF(sizeof(*eth))));
+
+	if (type != htons(ETH_P_XIP))
+		return -1;
+
+	eth->h_proto = type;
+	memcpy(eth->h_source, dev->dev_addr, ETH_ALEN);
+	memcpy(eth->h_dest, mfxid->ha, ETH_ALEN);
+	hh->hh_len = ETH_HLEN;
+	return 0;
+}
+
+/* Updates the cached hdr on change of address
+* type : 1 -> destionation addr
+*        0 -> source addr
+*/
+static void xia_ether_header_cache_update(struct hh_cache *hh,
+					  const struct net_device *dev,
+					  const unsigned char *haddr,
+					  const int type)
+{
+	struct ethhdr *eth;
+
+	eth = (struct ethhdr *)
+		    (((u8 *)hh->hh_data) + (HH_DATA_OFF(sizeof(*eth))));
+	if (type)
+		memcpy(eth->h_dest, haddr, ETH_ALEN);
+	else
+		memcpy(eth->h_source, dev->dev_addr, ETH_ALEN);
+}
+
+/* Cached Hdr functions. */
+static int mfxid_hh_init(struct fib_xid_ether_main *mfxid)
```

Comment:
Once the last two parameters of `xia_ether_header_cache()` are dropped, this function won't be doing much. In addition, it's only called in one place. So just replace that call of `mfxid_hh_init()` with `xia_ether_header_cache(mfxid)`.